controllers/argocd/notifications/deployment.go

```diff
@@ -0,0 +1,202 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func (nr *NotificationsReconciler) reconcileDeployment() error {
+
+	nr.Logger.Info("reconciling deployments")
+
+	notificationEnv := nr.Instance.Spec.Notifications.Env
+	notificationEnv = util.EnvMerge(notificationEnv, ProxyEnvVars(), false)
+
+	podSpec := corev1.PodSpec{
+		Volumes: []corev1.Volume{
+			{
+				Name: TLSCerts,
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: common.ArgoCDTLSCertsConfigMapName,
+						},
+					},
+				},
+			},
+			{
+				Name: ArgoCDRepoServerTLS,
+				VolumeSource: corev1.VolumeSource{
+					Secret: &corev1.SecretVolumeSource{
+						SecretName: common.ArgoCDRepoServerTLSSecretName,
+						Optional:   util.BoolPtr(true),
+					},
+				},
+			},
+		},
+		Containers: []corev1.Container{{
+			Command:         GetNotificationsCommand(nr.Instance),
+			Image:           GetArgoContainerImage(nr.Instance),
+			ImagePullPolicy: corev1.PullAlways,
+			Name:            common.ArgoCDNotificationsControllerComponent,
+			Env:             notificationEnv,
+			Resources:       GetNotificationsResources(nr.Instance),
+			LivenessProbe: &corev1.Probe{
+				ProbeHandler: corev1.ProbeHandler{
+					TCPSocket: &corev1.TCPSocketAction{
+						Port: intstr.IntOrString{
+							IntVal: int32(9001),
+						},
+					},
+				},
+			},
+			SecurityContext: &corev1.SecurityContext{
+				AllowPrivilegeEscalation: util.BoolPtr(false),
+				Capabilities: &corev1.Capabilities{
+					Drop: []corev1.Capability{
+						CapabilityDropAll,
+					},
+				},
+			},
+			VolumeMounts: []corev1.VolumeMount{
+				{
+					Name:      TLSCerts,
+					MountPath: VolumeMountPathTLS,
+				},
+				{
+					Name:      ArgoCDRepoServerTLS,
+					MountPath: VolumeMountPathRepoServerTLS,
+				},
+			},
+			WorkingDir: WorkingDirApp,
+		}},
+		SecurityContext: &corev1.PodSecurityContext{
+			RunAsNonRoot: util.BoolPtr(true),
+		},
+	}
+
+	AddSeccompProfileForOpenShift(nr.Logger, nr.Client, &podSpec)
+
+	deploymentRequest := workloads.DeploymentRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resourceName,
+			Namespace: nr.Instance.Namespace,
+			Labels:    resourceLabels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			},
+			Template: corev1.PodTemplateSpec{
+				Spec: podSpec,
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: resourceLabels,
+				},
+			},
+			Selector: &metav1.LabelSelector{
+				MatchLabels: resourceLabels,
+			},
+			Replicas: GetArgoCDNotificationsControllerReplicas(nr),
+		},
+
+		Client:    nr.Client,
+		Mutations: []mutation.MutateFunc{mutation.ApplyReconcilerMutation},
+	}
```

Comment:
I think we should extract the creation of desired deployment and deploymentRequest into 2 separate functions of their own and keep only the reconciliation of the deployment in `reconcileDeployment`. That will allow us to reuse the desired deployment spec in unit tests as well and reduce the size of the `reconcileDeployment` function